docs/architecture/adr-036-blockchain-reactor-refactor.md

```diff
@@ -0,0 +1,406 @@
+}
+
+type TerminationTimeout struct {
+	Height int64
+}
+
+type Message int
+
+const (
+	MessageUnknown Message = iota
+	MessageBlockRequest
+	MessageBlockResponse
+	MessageStatusReport
+)
+
+type BlockRequestMessage struct {
+	Height int64
+	PeerID ID
+}
+
+type BlockResponseMessage struct {
+	Height        int64
+	Block         Block
+	Commit        Commit
+	PeerID        ID
+	CurrentHeight int64
+}
+
+type StatusReportMessage struct {
+	Height int64
+}
+
+type TimeoutTrigger int
+
+const (
+	TimeoutUnknown TimeoutTrigger = iota
+	TimeoutResponseTrigger
+	TimeoutTerminationTrigger
+)
+```
+
+The Controller state machine can be in two modes (states): `FastSyncMode` when
+it is trying to catch up with the network by downloading committed blocks,
+and `ConsensusMode` in which it executes Tendermint consensus protocol. We
+assume that "fast sync" mode terminates once the Controller switch to
+`ConsensusMode`.  
+
+``` go
+type State int
+
+const (
+	StateUnknown State = iota
+	StateFastSyncMode
+	StateConsensusMode
+)
+
+type ControllerState struct {
+	Height             int64            // the first block that is not committed
+	State              State            // mode of operation of the state machine
+	PeerMap            map[ID]PeerStats // map of peer IDs to peer statistics
+	MaxRequestPending  int64            // maximum height of the pending requests
+	FailedRequests     []int64          // list of failed block requests
+	PendingRequestsNum int              // number of pending requests
+	Store              []BlockInfo      // contains list of downloaded blocks
+	Executor           BlockExecutor    // store, verify and executes blocks
+}
+
+type PeerStats struct {
+	Height         int64
+	PendingRequest int64 // it can be list in case there are multiple outstanding requests per peer
+}
+
+type BlockInfo struct {
+	Block  Block
+	Commit Commit
+	PeerID ID // a peer from which we downloading the corresponding Block and Commit
+}
+```
+
+The `Controller` is initialized by providing initial height (`startHeight`) from which it will start downloading blocks from peers.
+
+``` go
+func ControllerInit(state ControllerState, startHeight int64) ControllerState {
+	state.Height = startHeight
+	state.State = FastSyncMode
+	state.MaxRequestPending = startHeight - 1
+	state.PendingRequestsNum = 0
+}
+
+func ControllerHandle(event Event, state ControllerState) (ControllerState, Message, TimeoutTrigger, Error) {
+	msg = nil
+	timeout = nil
+	error = nil
+
+	switch state.State {
+	case StateConsensusMode:
+		switch event := event.(type) {
+		case EventBlockRequest:
+			msg = createBlockResponseMessage(state, event)
+			return state, msg, timeout, error
+		default:
+			error = "Only respond to BlockRequests while in ConsensusMode!"
+			return state, msg, timeout, error
+		}
+
+	case StateFastSyncMode:
+		switch event := event.(type) {
+		case EventBlockRequest:
+			msg = createBlockResponseMessage(state, event)
+			return state, msg, timeout, error
```

Comment:
Here we are not in consensus mode and we reply to Block response messages. Also, this (the fast sync helper functionality) could be done outside the controller.